### PR TITLE
Add MHD module and update testing

### PR DESCRIPTION
**Pull request by rfj82982**
This PR is to merge into main sources the MHD module. Testing has also been updated to include writing checkpoints at the end of simulation, test the averages collection and write two snapshots. 
Modifications have also been done to have compatibility with ADIOS2 IO
Some points still need to be addressed in future PR: 

- Naming of the variable in the checkpoint file with ADIOS2
- Fully include in CTest the Python to test the output file for the TGV case
- Divide the examples from the tests

This PR is addressing one of the requirements in issue #283 for the release of V5.0

### CODE REVIEW

File: examples/MHD/input_otv_test.i3d [hunk 1]
```diff
@@ -93,6 +93,27 @@ nstat = 1             ! Size arrays for statistic collection
 
 /End
 
+!====================
+&PartiParam
```
Comment (mathrack) [inline examples/MHD/input_otv_test.i3d]: The `PartiParam` section should be moved to a dedicated PR with the particles

File: src/Case-ABL.f90 [hunk 2]
```diff
@@ -120,20 +120,20 @@ subroutine init_abl(ux1,uy1,uz1,ep1,phi1)
...
-        !  do k=1,xsize(3)
-        !  do i=1,xsize(1)
-        !    call random_number(phinoise)
-        !    phinoise=0.1*(phinoise*2.-1.)
-        !    phi1(i,j,k,1)=phi1(i,j,k,1)+phinoise
-        !  enddo
-        !  enddo
-        !endif
-      enddo
+      !! Add random noise
```
Comment (mathrack) [inline src/Case-ABL.f90]: This could be removed ?

File: src/Case-Channel.f90 [hunk 3]
```diff
@@ -93,7 +93,7 @@ subroutine init_channel (ux1,uy1,uz1,ep1,phi1)
                 if (idir_stream == 1) then
                    ux1(i,j,k)=one-y*y
                    uy1(i,j,k)=zero
-                   uz1(i,j,k)=sin(real(i-1,mytype)*dx)+cos(real(k-1,mytype)*dz)
+                   uz1(i,j,k)=zero !sin(real(i-1,mytype)*dx)+cos(real(k-1,mytype)*dz)
```
Comment (mathrack) [inline src/Case-Channel.f90]: This was probably present to help the transition to turbulence ?

I suggest to remove the commented code or at the very least to replace `k-1` with `k+xstart(3)-2`

File: src/mhd.f90 [hunk 4]
```diff
@@ -165,89 +189,89 @@ end function vortcal
...
-    endif
-    !
-    Ek=0._mytype
-    Em=0._mytype
-    Omegam=0._mytype
-    Jmax=0._mytype
-    do k=1,xsize(3)
-    do j=1,xsize(2)
-    do i=1,xsize(1)
+  ! subroutine mhd_sta(ux1,uy1,uz1)
```
Comment (mathrack) [inline src/mhd.f90]: Is it needed to keep a lot of commented code ?

File: src/mhd.f90 [hunk 5]
```diff
@@ -883,8 +909,23 @@ function divergence_scalar(vec,nlock) result(pp3)
     if ((nrank == 0) .and. (nlock > 0).and.(mod(itime, ilist) == 0 .or. itime == ifirst .or. itime==ilast)) then
        if (nlock == 2) then
           write(*,*) 'DIV B  max mean=',real(tmax1,mytype),real(tmoy1/real(nproc),mytype)
+          ! if(itime==6000) then
```
Comment (mathrack) [inline src/mhd.f90]: Is it needed to keep a lot of commented code ?

File: src/navier.f90 [hunk 6]
```diff
@@ -621,9 +621,9 @@ subroutine pre_correc(ux,uy,uz,ep)
           enddo
           do k=1,xsize(3)
              do i=1,xsize(1)
-                ux(i,1,k)=byx1(i,k)+dpdxy1(i,k)
+                ux(i,1,k)=byx1(i,k) !+dpdxy1(i,k)
```
Comment (mathrack) [inline src/navier.f90]: This looks like a big modification ?

File: src/xcompact3d.f90 [hunk 7]
```diff
@@ -273,11 +272,29 @@ subroutine init_xcompact3d()
   if (itype==2) then
      if(nrank.eq.0)then
         open(42,file='time_evol.dat',form='formatted')
+        if(mhd_active) then
+           open(43,file='mhd_time_evol.dat',form='formatted')
```
Comment (mathrack) [inline src/xcompact3d.f90:277]: Hard-coded IO unit number is not looking great ! But if it works just fine, ok.

File: src/xcompact3d.f90 [hunk 8]
```diff
@@ -273,11 +272,29 @@ subroutine init_xcompact3d()
...
+     endif
+  endif
+
+  if (itype==3) then
+     if (nrank.eq.0)then
+        open(52,file='chan_time_evol.dat',form='formatted')
+     endif
+  endif
+  
+  if (itype==5) then
```
Comment (mathrack) [inline src/xcompact3d.f90]: This is not equivalent to the current code.

File: src/Case-Channel.f90 [hunk 9]
```diff
@@ -274,19 +260,111 @@ subroutine channel_cfr (ux, constant)
...
+             do i=1,xsize1
+              temp2=max(temp2,ux1(i,j,k))
+             enddo
+          enddo
+       enddo
+       
+       call MPI_ALLREDUCE(temp1,ufriction,1,real_type,MPI_SUM,MPI_COMM_WORLD,code)
+       call MPI_ALLREDUCE(temp2,     umax,1,real_type,MPI_MAX,MPI_COMM_WORLD,code)
+
+       ufriction=xnu*1.5_mytype*ufriction/(nxc*nzc) ! ub=2/3
```
Comment (mathrack) [inline src/Case-Channel.f90]: In my opinion, this is not the best way to monitor the friction

**Review by mathrack — COMMENTED** [review]
The PR is not easy to review because it goes beyond the simple addition of the MHD module.

File: src/Case-Channel.f90 [hunk 10]
```diff
@@ -274,19 +260,111 @@ subroutine channel_cfr (ux, constant)
...
+
+       endif
+       !
+       if (nclyn==2 .and. xend(2)==ny) then
+
+         ! upper wall, one node away from wall
+         j=xsize(2)-1
+         do k=1,xsize(3)
+             do i=1,xsize(1)
+              temp1=temp1-ux1(i,j,k)*ppy(j+1)
```
Comment (mathrack) [inline src/Case-Channel.f90]: `ppy(ny)`

File: src/Case-Channel.f90 [hunk 11]
```diff
@@ -274,19 +260,111 @@ subroutine channel_cfr (ux, constant)
...
+       temp2=zero
+       do k=1,xsize3
+          do j=1,xsize2
+             do i=1,xsize1
+              temp2=max(temp2,ux1(i,j,k))
+             enddo
+          enddo
+       enddo
+       
+       call MPI_ALLREDUCE(temp1,ufriction,1,real_type,MPI_SUM,MPI_COMM_WORLD,code)
```
Comment (mathrack) [inline src/Case-Channel.f90]: temporary variables can be avoided when performing MPI reduction in place

File: src/visu.f90 [hunk 12]
```diff
@@ -96,7 +97,9 @@ subroutine visu_init()
     endif
     if (iscalar.ne.0) then
        do is = 1, numscalar
-          call decomp_2d_register_variable(io_name, "phi"//char(48+is), 1, 0, output2D, mytype)
+          write(scname,"('phi',I2.2)") is
+          !call decomp_2d_register_variable(io_name, "phi"//char(48+is), 1, 0, output2D, mytype)
```
Comment (mathrack) [inline src/visu.f90:101]: The commented code should be removed

File: examples/MHD/input_otv_test.i3d (hunk 1 above)
Comment (rfj82982) [reply examples/MHD/input_otv_test.i3d]: Fixed this. It was a refuse from the particle development

File: src/Case-ABL.f90 (hunk 2 above)
Comment (rfj82982) [reply src/Case-ABL.f90]: No idea if useful or not. We should ask to someone that uses the ABL case to see if the noise for Phi is necessary or not

File: src/navier.f90 (hunk 6 above)
Comment (rfj82982) [reply src/navier.f90]: I need to check this with Jian. I didn't spot this before.

File: src/xcompact3d.f90 (hunk 7 above)
Comment (rfj82982) [reply src/xcompact3d.f90:277]: The hard-coded IO unit has always been there. I am ok to change it, but I would do it in a dedicated PR

File: src/Case-Channel.f90 (hunk 9 above)
Comment (rfj82982) [reply src/Case-Channel.f90]: Need to check this with Jian

File: examples/MHD/input_otv.i3d [hunk 13]
```diff
@@ -8,15 +8,15 @@
...
-p_row=0               ! Row partition
-p_col=0               ! Column partition
-
-! Mesh
-nx=128                ! X-direction nodes
-ny=128                ! Y-direction nodes
-nz=1                  ! Z-direction nodes
-istret = 0            ! y mesh refinement (0:no, 1:center, 2:both sides, 3:bottom)
-beta = 0.259065151    ! Refinement parameter (beta)
+p_row=2                ! Row partition
```
Comment (mathrack) [inline examples/MHD/input_otv.i3d]: Keep `p_row` and `p_col` at zero ? Currently, the test can run only on 4 CPUs

File: examples/MHD/input_otv.i3d [hunk 14]
```diff
@@ -93,6 +93,27 @@ nstat = 1             ! Size arrays for statistic collection
 
 /End
 
+!====================
+&PartiParam
```
Comment (mathrack) [inline examples/MHD/input_otv.i3d]: The `PartiParam` section should be moved to a dedicated PR with the particles

File: src/Case-TGV.f90 [hunk 15]
```diff
@@ -412,6 +375,101 @@ subroutine postprocess_tgv(ux1,uy1,uz1,phi1,ep1)
...
+          temp3=zero
+          do k=1,xsize3
+             do j=1,xsize2
+                do i=1,xsize1
+                   temp1=temp1+zpfive*((Bm(i,j,k,1))**2+(Bm(i,j,k,2))**2+(Bm(i,j,k,3))**2)
+                   temp2=temp2+zpfive*Je(i,j,k,1)**2+Je(i,j,k,2)**2+Je(i,j,k,3)**2
+                enddo
+             enddo
+          enddo
+          call MPI_ALLREDUCE(temp1,   eem,1,real_type,MPI_SUM,MPI_COMM_WORLD,code)
```
Comment (mathrack) [inline src/Case-TGV.f90:393]: temporary variables can be avoided when performing MPI reduction in place

File: src/Case-TGV.f90 [hunk 16]
```diff
@@ -412,6 +375,101 @@ subroutine postprocess_tgv(ux1,uy1,uz1,phi1,ep1)
...
+                                         (two*ti1(i,j,k))**two+two*(td1(i,j,k)+tb1(i,j,k))**two+&
+                                                               two*(tg1(i,j,k)+tc1(i,j,k))**two+&
+                                                               two*(th1(i,j,k)+tf1(i,j,k))**two)
+                   temp2=temp2+zpfive*((tf1(i,j,k)-th1(i,j,k))**2+&
+                                       (tg1(i,j,k)-tc1(i,j,k))**2+&
+                                       (tb1(i,j,k)-td1(i,j,k))**2)
+                enddo
+             enddo
+          enddo
+          call MPI_ALLREDUCE(temp1,disb,1,real_type,MPI_SUM,MPI_COMM_WORLD,code)
```
Comment (mathrack) [inline src/Case-TGV.f90:460]: temporary variables can be avoided when performing MPI reduction in place

File: src/mhd.f90 (hunk 5 above)
Comment (rfj82982) [reply src/mhd.f90]: This has been removed

File: examples/MHD/input_otv.i3d (hunk 13 above)
Comment (slaizet) [reply examples/MHD/input_otv.i3d]: agreed, this needs to be changed

File: examples/MHD/input_otv.i3d [hunk 17]
```diff
@@ -93,6 +93,16 @@ nstat = 1             ! Size arrays for statistic collection
 
 /End
 
+!====================
+&MHDParam
+!====================
+
+mhd_equation = 'induction'
+Stuart = 17.54386
```
Comment (slaizet) [inline examples/MHD/input_otv.i3d]: provide definition for all the different Reynolds numbers

File: src/Case-ABL.f90 (hunk 2 above)
Comment (slaizet) [reply src/Case-ABL.f90]: Why this part of the code has been changed? This is not related to MHD

File: src/Case-Channel.f90 (hunk 3 above)
Comment (slaizet) [reply src/Case-Channel.f90]: This should not have been commented

File: src/Case-Channel.f90 (hunk 10 above)
Comment (slaizet) [reply src/Case-Channel.f90]: to be fixed?

File: src/Case-Channel.f90 (hunk 11 above)
Comment (slaizet) [reply src/Case-Channel.f90]: in the interest of time, let's keep this part as it is

File: src/Case-Channel.f90 (hunk 9 above)
Comment (slaizet) [reply src/Case-Channel.f90]: Agreed with Cedric

File: src/Case-TGV.f90 [hunk 18]
```diff
@@ -67,17 +68,24 @@ subroutine init_tgv (ux1,uy1,uz1,ep1,phi1)
...
-                   Bm(i,j,k,3)=zero
+                   ! OTV
+                   ux1(i,j,k)=-two*sin(y)
+                   uy1(i,j,k)= two*sin(x)
+                   uz1(i,j,k)= zero
                    !
-                   Bmean(i,j,k,1)=zero
-                   Bmean(i,j,k,2)=zero
-                   Bmean(i,j,k,3)=zero
+                   if(nz<=2) then
```
Comment (slaizet) [inline src/Case-TGV.f90:77]: why "nz<=2"?

File: src/Case-TGV.f90 (hunk 15 above)
Comment (slaizet) [reply src/Case-TGV.f90:393]: let's not worry about that now

File: src/Case-Channel.f90 (hunk 9 above)
Comment (rfj82982) [reply src/Case-Channel.f90]: If you agree I'll completely remove the monitoring of the friction

File: src/navier.f90 (hunk 6 above)
Comment (slaizet) [reply src/navier.f90]: there should be a if statement here:
-keep the pressure gradient correction if no MHD, remove them if MHD

File: examples/MHD/input_otv.i3d (hunk 14 above)
Comment (rfj82982) [reply examples/MHD/input_otv.i3d]: Resolved

File: src/Case-Channel.f90 (hunk 9 above)
Comment (slaizet) [reply src/Case-Channel.f90]: yes please

File: src/Case-Channel.f90 (hunk 3 above)
Comment (mathrack) [reply src/Case-Channel.f90]: The `k-1` was incorrect. We can ship a fix in this PR or in a dedicated one.

File: src/Case-ABL.f90 (hunk 2 above)
Comment (rfj82982) [reply src/Case-ABL.f90]: I have removed the random noise since it was not used. However, just to be clear this was already here before and it has not been added with the MHD module

File: src/Case-Channel.f90 (hunk 3 above)
Comment (rfj82982) [reply src/Case-Channel.f90]: I have added @mathrack correction for the channel in X. Same should be done also for Z but i think we can do this in a dedicated PR, after testing all channel and relative initialisation.

File: src/navier.f90 (hunk 6 above)
Comment (rfj82982) [reply src/navier.f90]: Ok I have split the no correction only for MHD and add the correction back in the no MHD case.

File: src/xcompact3d.f90 (hunk 8 above)
Comment (rfj82982) [reply src/xcompact3d.f90]: I have removed the `itype==5` if statement

File: src/Case-Channel.f90 (hunk 9 above)
Comment (rfj82982) [reply src/Case-Channel.f90]: Ok removed all code in postprocess_channel

File: src/Case-Channel.f90 (hunk 11 above)
Comment (rfj82982) [reply src/Case-Channel.f90]: Ok I'll change all with in place is a future PR

File: src/visu.f90 (hunk 12 above)
Comment (rfj82982) [reply src/visu.f90:101]: I'll keep this comment for the moment just to have a reference to the olf way of creating the variable name. This is until we'll fix the issue with ADIOS2.

File: src/Case-TGV.f90 (hunk 15 above)
Comment (rfj82982) [reply src/Case-TGV.f90:393]: As comment before

File: src/Case-TGV.f90 (hunk 18 above)
Comment (rfj82982) [reply src/Case-TGV.f90:77]: I guess this is to distinguish between 3D and 2D cases.

File: examples/MHD/input_otv.i3d (hunk 13 above)
Comment (rfj82982) [reply examples/MHD/input_otv.i3d]: resolved

File: src/Case-TGV.f90 (hunk 18 above)
Comment (rfj82982) [reply src/Case-TGV.f90:77]: Yes initialisation is different between 2D and 3D. I guess this is OK and the only way to identify a 2D case.

File: examples/MHD/input_otv.i3d (hunk 17 above)
Comment (rfj82982) [reply examples/MHD/input_otv.i3d]: Resolved

File: src/Case-Channel.f90 (hunk 10 above)
Comment (rfj82982) [reply src/Case-Channel.f90]: This has been removed since was part of the postprocessing
